### PR TITLE
Make the batch_idx argument optional in all step methods

### LINKED ISSUE #18506: Make the batch_idx argument optional in all step methods
### Bug description

The training_step method does not require declaring a batch_idx argument:

```py
def training_step(self, batch, batch_idx):
    pass

# or
def training_step(self, batch):
    pass
```

But for validation/test/predict steps, the following raises an error:

```py
def validation_step(self, batch):
    pass
```

Is this intentional or a bug? 

### What version are you seeing the problem on?

v2.0, master

### How to reproduce the bug

```python
import os

import torch
from torch.utils.data import DataLoader, Dataset

from lightning.pytorch import LightningModule, Trainer


class RandomDataset(Dataset):
    def __init__(self, size, length):
        self.len = length
        self.data = torch.randn(length, size)

    def __getitem__(self, index):
        return self.data[index]

    def __len__(self):
        return self.len


class BoringModel(LightningModule):
    def __init__(self):
        super().__init__()
        self.layer = torch.nn.Linear(32, 2)

    def forward(self, x):
        return self.layer(x)

    def training_step(self, batch, batch_idx):
        return self(batch).sum()

    def validation_step(self, batch):
        self(batch).sum()

    def configure_optimizers(self):
        return torch.optim.SGD(self.layer.parameters(), lr=0.1)


def run():
    train_data = DataLoader(RandomDataset(32, 64), batch_size=2)
    val_data = DataLoader(RandomDataset(32, 64), batch_size=2)

    model = BoringModel()
    trainer = Trainer(
        default_root_dir=os.getcwd(),
        limit_train_batches=1,
        limit_val_batches=1,
        limit_test_batches=1,
        num_sanity_val_steps=0,
        max_epochs=1,
        enable_model_summary=False,
    )
    trainer.fit(model, train_dataloaders=train_data, val_dataloaders=val_data)


if __name__ == "__main__":
    run()
```

### Error messages and logs

```
  File "/Users/adrian/repositories/lightning/examples/pytorch/bug_report/bug_report_model.py", line 57, in <module>
    run()
  File "/Users/adrian/repositories/lightning/examples/pytorch/bug_report/bug_report_model.py", line 53, in run
    trainer.fit(model, train_dataloaders=train_data, val_dataloaders=val_data)
  File "/Users/adrian/repositories/lightning/src/lightning/pytorch/trainer/trainer.py", line 542, in fit
    call._call_and_handle_interrupt(
  File "/Users/adrian/repositories/lightning/src/lightning/pytorch/trainer/call.py", line 44, in _call_and_handle_interrupt
    return trainer_fn(*args, **kwargs)
  File "/Users/adrian/repositories/lightning/src/lightning/pytorch/trainer/trainer.py", line 581, in _fit_impl
    self._run(model, ckpt_path=ckpt_path)
  File "/Users/adrian/repositories/lightning/src/lightning/pytorch/trainer/trainer.py", line 990, in _run
    results = self._run_stage()
  File "/Users/adrian/repositories/lightning/src/lightning/pytorch/trainer/trainer.py", line 1033, in _run_stage
    self.fit_loop.run()
  File "/Users/adrian/repositories/lightning/src/lightning/pytorch/loops/fit_loop.py", line 202, in run
    self.advance()
  File "/Users/adrian/repositories/lightning/src/lightning/pytorch/loops/fit_loop.py", line 363, in advance
    self.epoch_loop.run(self._data_fetcher)
  File "/Users/adrian/repositories/lightning/src/lightning/pytorch/loops/training_epoch_loop.py", line 137, in run
    self.on_advance_end(data_fetcher)
  File "/Users/adrian/repositories/lightning/src/lightning/pytorch/loops/training_epoch_loop.py", line 280, in on_advance_end
    self.val_loop.run()
  File "/Users/adrian/repositories/lightning/src/lightning/pytorch/loops/utilities.py", line 181, in _decorator
    return loop_run(self, *args, **kwargs)
  File "/Users/adrian/repositories/lightning/src/lightning/pytorch/loops/evaluation_loop.py", line 137, in run
    self._evaluation_step(batch, batch_idx, dataloader_idx, dataloader_iter)
  File "/Users/adrian/repositories/lightning/src/lightning/pytorch/loops/evaluation_loop.py", line 391, in _evaluation_step
    output = call._call_strategy_hook(trainer, hook_name, *step_args)
  File "/Users/adrian/repositories/lightning/src/lightning/pytorch/trainer/call.py", line 309, in _call_strategy_hook
    output = fn(*args, **kwargs)
  File "/Users/adrian/repositories/lightning/src/lightning/pytorch/strategies/strategy.py", line 401, in validation_step
    return self.lightning_module.validation_step(*args, **kwargs)
TypeError: BoringModel.validation_step() takes 2 positional arguments but 3 were given
```

### Environment

<details>
  <summary>Current environment</summary>

```
#- Lightning Component (e.g. Trainer, LightningModule, LightningApp, LightningWork, LightningFlow):
#- PyTorch Lightning Version (e.g., 1.5.0): 2.1.0dev
#- Lightning App Version (e.g., 0.5.2): -
#- PyTorch Version (e.g., 2.0): 2.0.1
#- Python version (e.g., 3.9): 3.10
#- OS (e.g., Linux): MacOs
#- CUDA/cuDNN version: -
#- GPU models and configuration: -
#- How you installed Lightning(`conda`, `pip`, source): source
#- Running environment of LightningApp (e.g. local, cloud):
```

</details>

### More info

_No response_

cc @carmocca @justusschock @awaelchli @borda

### PR DESCRIPTION
**Pull request by 0x404**
## What does this PR do?

make `batch_idx` argument optional in `test_step`, `validation_step` and `predict_step`.

Fixes #18506

<details>
  <summary><b>Before submitting</b></summary>

- Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- Did you make sure to **update the documentation** with your changes? (if necessary)
- Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- Did you list all the **breaking changes** introduced by this pull request?
- Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

----
:books: Documentation preview :books:: https://pytorch-lightning--18512.org.readthedocs.build/en/18512/

### CODE REVIEW

File: tests/tests_pytorch/loops/test_evaluation_loop.py [hunk 1]
```diff
@@ -496,3 +496,28 @@ def test_step(self, batch, batch_idx):
...
+            return {"x": self.step(batch)}
+
+        def test_step(self, batch):
+            self.test_step_called = True
+            return {"y": self.step(batch)}
+
+    trainer = Trainer(default_root_dir=tmpdir, fast_dev_run=1)
+    model = TestModel()
+
+    trainer.fit(model)
```
Comment (carmocca) [inline tests/tests_pytorch/loops/test_evaluation_loop.py:519]:
```suggestion
    trainer.validate(model)
```

**Review by carmocca — APPROVED** [review]
LGTM!

**Review by awaelchli — APPROVED** [review]

File: tests/tests_pytorch/loops/test_evaluation_loop.py [hunk 2]
```diff
@@ -496,3 +496,28 @@ def test_step(self, batch, batch_idx):
...
+
+        def validation_step(self, batch):
+            self.validation_step_called = True
+            return {"x": self.step(batch)}
+
+        def test_step(self, batch):
+            self.test_step_called = True
+            return {"y": self.step(batch)}
+
+    trainer = Trainer(default_root_dir=tmpdir, fast_dev_run=1)
```
Comment (awaelchli) [inline tests/tests_pytorch/loops/test_evaluation_loop.py:516]: We should probably set `logger=False`, `enable_checkpointing=False`, `enable_progress_bar=False` in these tests

File: src/lightning/pytorch/CHANGELOG.md [hunk 3]
```diff
@@ -185,6 +185,9 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 - Due to lack of reliability, Trainer now only runs on one GPU instead of all GPUs in a Jupyter notebook if `devices="auto"` (default) ([#18291](https://github.com/Lightning-AI/lightning/pull/18291))
 
 
+- make the `batch_idx` argument optional in `validation_step`, `test_step` and `predict_step` to maintain consistency with `training_step` ([#18512](https://github.com/Lightning-AI/lightning/pull/18512))
```
Comment (awaelchli) [inline src/lightning/pytorch/CHANGELOG.md:188]:
```suggestion
- Made the `batch_idx` argument optional in `validation_step`, `test_step` and `predict_step` to maintain consistency with `training_step` ([#18512](https://github.com/Lightning-AI/lightning/pull/18512))
```

File: tests/tests_pytorch/loops/test_evaluation_loop.py (hunk 2 above)
Comment (0x404) [reply tests/tests_pytorch/loops/test_evaluation_loop.py]: > Revised, I feel the reason for doing this is to reduce the execution of unrelated code during testing. Are there any other reasons why we need to do this?

Comment (carmocca) [reply tests/tests_pytorch/loops/test_evaluation_loop.py]: That's right. It'll make the test run faster too

**Pull request merged**